src/platform/src/main/java/org/geoserver/platform/resource/FileSystemResourceStore.java

```diff
@@ -476,15 +478,24 @@ public boolean delete() {
 
         @Override
         public boolean renameTo(Resource dest) {
-            if(dest instanceof FileSystemResource) {
-                return file.renameTo(((FileSystemResource)dest).file);
-            } else if(dest instanceof Files.ResourceAdaptor) {
-                    return file.renameTo(((Files.ResourceAdaptor)dest).file);
-            } else {
-                return Resources.renameByCopy(this, dest);
+            if (dest.parent().path().contains(path())) {
+                LOGGER.log(Level.FINE, "Cannot rename a resource to a descendant of itself");
+                return false;
             }
+            try {
+                if(dest instanceof FileSystemResource) {
+                    rename(file, ((FileSystemResource)dest).file);
+                } else if(dest instanceof Files.ResourceAdaptor) {
+                    rename(file, ((Files.ResourceAdaptor)dest).file);
```

Comment:
Interesting we have to special case the adaptor, I am not sure how often these are used/useful.